packages/runtime/src/state/react-builder-preview.ts

```diff
@@ -402,15 +404,50 @@ function startPollingElementFromPoint(): ThunkAction<() => void, State, unknown,
   }
 }
 
-export function initialize(): ThunkAction<() => void, State, unknown, Action> {
-  return dispatch => {
+export interface IMessageChannel {
+  postMessage(message: any, transferables?: Transferable[]): void
+  dispatchBuffered(): void
+}
+
+function registerBuilderComponents(state: State, messageChannel: IMessageChannel) {
+  const componentsMeta = getComponentsMeta(state)
+  componentsMeta.forEach((componentMeta, componentType) => {
+    const propControllerDescriptors = getComponentPropControllerDescriptors(state, componentType)
+
+    if (propControllerDescriptors != null) {
+      const [serializedControls, transferables] = serializeControls(propControllerDescriptors)
+
+      messageChannel.postMessage(
+        registerBuilderComponent(componentType, componentMeta, serializedControls),
+        transferables,
+      )
+    }
+  })
+
+  return () => {
+    componentsMeta.forEach((_, componentType) => {
+      messageChannel.postMessage(unregisterBuilderComponent(componentType))
+    })
+  }
+}
+
+export function initialize(
+  channel: IMessageChannel,
+): ThunkAction<() => void, State, unknown, Action> {
+  return (dispatch, getState) => {
+    const state = getState()
     const stopMeasuringElements = dispatch(startMeasuringElements())
     const stopMeasuringDocumentElement = dispatch(startMeasuringDocumentElement())
     const stopHandlingFocusEvent = dispatch(startHandlingFocusEvents())
     const unlockDocumentScroll = dispatch(lockDocumentScroll())
     const stopHandlingPointerMoveEvent = dispatch(startHandlingPointerMoveEvent())
     const stopPollingElementFromPoint = dispatch(startPollingElementFromPoint())
+    const unregisterBuilderComponents = registerBuilderComponents(state, channel)
+
+    const breakpoints = ReactPage.getBreakpoints(state)
+    dispatch(setBreakpoints(breakpoints))
     dispatch(setIsInBuilder(true))
+    channel.dispatchBuffered()
```

Comment:
Dispatch channel's buffered messages, if any, as a part of the initialization sequence after the connection has been established.